run.c

```diff
@@ -399,6 +432,20 @@ void bpe_encode(char *text, char **vocab, float *vocab_scores, int vocab_size, u
     }
 
     free(str_buffer);
+    free(sorted_vocab);
+}
+
+// convert token to printable string
+char *token_to_str(char **vocab, int token, int prev_token) {
+    // following BOS (1) token, sentencepiece decoder strips any leading whitespace (see PR #89)
+    char *token_str = (prev_token == 1 && vocab[token][0] == ' ') ? vocab[token]+1 : vocab[token];
+    // make '<0x01>' into '\x01'
+    static char byte_piece[4];
+    if (sscanf(token_str, "<0x%02X>", (int*)(&byte_piece)) == 1) {
```

Comment:
is this part necessary? maybe we could just leave the strings like <0x01> as they are?

Reply:
Depends on basic alphabet coverage in vocabulary. I can check llama2 tokenizer vocabulary for letters/languages I know but can't comment for other languages. 

if we remove those, we can also remove byte_fallback encoding too, and error out if token is not in vocabulary as before?
```
            // byte_fallback encoding
            for (int i=0; i<str_len; i++) {
                tokens[(*n_tokens)++] = (unsigned char)str_buffer[i] + 3;
            }
```

Reply:
if i remember correctly sentencepiece maintains decode(encode(normalize(s))) == normalize(s) property. decode() will output control characters as is if it was in original string/token.

Reply:
also there is iscntrl() C function but looks like '\n' is there too, and we probably want to print newlines as is.

Reply:
update: 10 languages I tested from lorem ipsum text generator (https://generator.lorem-ipsum.info/), only 3 did not require byte_fallback when encoding 1st short paragraph.